Use CLOCK_MONOTONIC

dofmind opened the pull request:
Relates to: https://github.com/eclipse-bluechi/bluechi/issues/952

CLOCK_REALTIME reflects the current date and time, which can be adjusted by the system (e.g. NTP synchronization or manual change by the administrator).

We get the current time to measue elapsed time, so that should use CLOCK_MONOTONIC instead of CLOCK_REALTIME.

So, this adds to use monotonic time for heartbeat and DBus properties for monotonic time.
Also get_time_seconds() is removed, as it can be replaced with get_time_micros() for precision.

42dot-joonyoung-shim committed:
Use monotonic time

This adds to use monotonic time for heartbeat and DBus properties for
monotonic time.

Also get_time_seconds() is removed, as it can be replaced with
get_time_micros() for precision.

Signed-off-by: Joonyoung Shim <joonyoung.shim@42dot.ai>

mkemel commented on src/agent/agent.c:
```diff
@@ -2313,14 +2307,23 @@ static int agent_match_job_removed(sd_bus_message *m, void *userdata, UNUSED sd_
         }
 
+        now_monotonic = get_time_micros_monotonic();
+        if (now == 0) {
```
```suggestion
        if (now_monotonic == 0) {
```

dofmind replied:
Oops, i will fix.

engelmi commented on src/libbluechi/common/time-util.c:
```diff
@@ -20,18 +20,14 @@ uint64_t get_time_micros() {
-        *ret_time_seconds = now.tv_sec;
-        return 0;
+        uint64_t now_micros = now.tv_sec * sec_to_microsec_multiplier +
+                        (uint64_t) ((double) now.tv_nsec * nanosec_to_microsec_multiplier);
```
Please move this calculation to a dedicated function so we can reuse it here and in `get_time_micros`. Would be great to have a simple unit test with a few cases as well.

dofmind replied:
OK, i will do it.

dofmind replied:
I added the timespec_to_micros function and an unit test for it.

dofmind replied:
Need to handle build errors on 32-bit system.

engelmi replied:
Thanks! I'll have another look.

engelmi reviewed (commented):
Small comment, otherwise looks good to me. 
Thanks for converting some of the property functions to simple `offsetof`!

42dot-joonyoung-shim committed:
Add timespec_to_micros in time-util

The function timespec_to_micros converts the time stored in a struct
timespec into microseconds (uint64_t type).

Signed-off-by: Joonyoung Shim <joonyoung.shim@42dot.ai>

mkemel reviewed (approved):
LGTM

engelmi commented on src/libbluechi/common/time-util.c:
```diff
@@ -15,23 +15,15 @@ uint64_t get_time_micros() {
-                return -EINVAL;
-        }
-
+uint64_t get_time_micros_monotonic() {
```
I'd prefer to keep the (systemd-like) style of returning the error code and using the parameter for returning. This way, the error logs where this function is called provides more information. 
```suggestion
int get_time_micros_monotonic(time_t *ret_time_seconds) {
```

The same applies for `timespec_to_micros` - maybe using `EINVAL` as return code?!
```
int timespec_to_micros(struct timespec time, uint64_t *ret){
   ...
   if (sec_micros > UINT64_MAX - nsec_micros) {
      return -EINVAL;
   }
   ...
```

dofmind replied:
If you perfer systemd-like style, what do you think about reusing code from time-util.c of systemd? It uses assert() for error cases and returns USEC_INFINITY(maximum value) for overflow or abnormal values.

e.g. systemd src/basic/time-util.c
```
usec_t now(clockid_t clock_id) {
        struct timespec ts;

        assert_se(clock_gettime(map_clock_id(clock_id), &ts) == 0);

        return timespec_load(&ts);
}
...
usec_t timespec_load(const struct timespec *ts) {
        assert(ts);

        if (ts->tv_sec < 0 || ts->tv_nsec < 0)
                return USEC_INFINITY;

        if ((usec_t) ts->tv_sec > (UINT64_MAX - (ts->tv_nsec / NSEC_PER_USEC)) / USEC_PER_SEC)
                return USEC_INFINITY;

        return
                (usec_t) ts->tv_sec * USEC_PER_SEC +
                (usec_t) ts->tv_nsec / NSEC_PER_USEC;
}
```

engelmi replied:
Makes sense to me. 
I probably wouldn't import `assert_se` from systemd, but use a plain `assert` when `clock_gettime` fails - that operation should never fail. Returning `USEC_INFINITY` in overflow/abnormal value cases is a good idea, I think. Could you define `USEC_INFINITY` in `time-util.h`? We have constants like `USEC_PER_SEC` already in [protocol.h](https://github.com/eclipse-bluechi/bluechi/blob/main/src/libbluechi/common/protocol.h#L10-L13), but to me that seems off now. I'd move these to `time-util.h` in a later PR.

dofmind replied:
Updated.

engelmi replied:
Thanks, looks good!

engelmi commented on src/libbluechi/common/time-util.c:
```diff
@@ -42,6 +34,23 @@ double micros_to_millis(uint64_t time_micros) {
+        uint64_t sec_micros = 0;
+        uint64_t nsec_micros = 0;
+
+        if ((uint64_t) time.tv_sec > UINT64_MAX / sec_to_microsec_multiplier) {
```
Not sure if these checks for the maximum are really necessary since the actual times will likely be much lower than what `uint64_t` can store. 
What do you think?

dofmind replied:
I think overflow checking is a good habit to have to prepare for unusual situations. There is already similar code in systemd.

engelmi reviewed (approved):
LGTM
